### ISSUE TITLE
Decal depth sampling broken when using MSAA (Workaround applied).

### ISSUE THREAD

**Issue by klorfmorf** [rt64/rt64#24]
First of all, great work on RT64, I knew HLE could be done better. 

Here are some screenshots showing the issue:

**Vulkan: MSAA 4x**
![Vulkan MSAA 4x](https://github.com/rt64/rt64/assets/169589138/d31becb6-4cdf-4667-a4a6-1100e43c5fb3)

**Vulkan: MSAA 2x**
![Vulkan MSAA 2x](https://github.com/rt64/rt64/assets/169589138/0a75b64d-1869-4d4e-b0a9-9553c12a4f66)

**Vulkan: MSAA Off**
![Vulkan MSAA Off](https://github.com/rt64/rt64/assets/169589138/7736c64d-b329-4e9d-b241-1410fa2c51ae)

**D3D12: MSAA 4x**
![D3D12 MSAA 4x](https://github.com/rt64/rt64/assets/169589138/323c03ae-7214-4765-8677-7ed90f3a96e8)

**D3D12: MSAA 2x**
![D3D12 MSAA 2x](https://github.com/rt64/rt64/assets/169589138/9cbe5dbd-b8ad-4d2a-8665-a25b54b22bde)

**D3D12: MSAA Off**
![D3D12 MSAA Off](https://github.com/rt64/rt64/assets/169589138/5f63f5cf-ba4f-4140-8df0-36b726a6f428)

As you can see, the issue is more a lot more visible on Vulkan than it is on D3D12.

Basically all of the shadows seem to exhibit this problem (also happens in water when the foam is around the player).

I remember GLideN64 also having a problem like this where "accurate depth comparison" could not be enabled with MSAA, however not  having that option enabled would lead to every shadow being rendered wrong (_for example, when standing near the edges of a floor, the shadows would render over the edges of the floor, making them float in air_).

There is another problem which is sometimes visible in D3D12, the half transparent shadow which can be seen in Vulkan with MSAA 2x can also be sometimes seen in any of the D3D12 modes where MSAA is on.

For reference, I am using an NVIDIA RTX 3060 with the latest drivers.

**Comment by klorfmorf** [comment]
Here are some screenshots showing the transparent shadow issue on D3D12:

**D3D12: MSAA 4x**
![D3D12 MSAA 4x Transparent Shadow](https://github.com/rt64/rt64/assets/169589138/1337220f-ff13-479f-8f2b-bdfbe2af08a6)

**D3D12: MSAA 2x**
![D3D12 MSAA 2x Transparent Shadow](https://github.com/rt64/rt64/assets/169589138/0e3f5760-326c-4fbe-a411-d10b8b2baca8)

**D3D12: MSAA Off*
![D3D12 MSAA Off Transparent Shadow](https://github.com/rt64/rt64/assets/169589138/4be27fed-186d-4698-916b-f419bd9f1439)

The shadow disappears on the ground around this area even on the Nintendo 64 (_which is emulated correctly with MSAA Off_), but the MSAA seems to add some kind of "ghosting" artifacts.

**Comment by DarioSamo** [comment]
Hey, thanks for the report!

I think I've seen this issue in Mario Tennis as well, it seems the other pixels of the multisampling don't seem to accurately sample the depth buffer for decals some reason when MSAA is used, but it doesn't seem like it happens consistently across games.

When looking at the resolve in RenderDoc it seems like basically every sample but the first one emulates the decal correctly. As far as I can tell the logic is correct, but it seems it doesn't like something about what I'm doing. I'm not really aware of what the reason behind the bug is yet so more in-depth investigation will be required.

(Casually dropping the Goemon recomp I see)

**Comment by DarioSamo** [comment]
> There is another problem which is sometimes visible in D3D12, the half transparent shadow which can be seen in Vulkan with MSAA 2x can also be sometimes seen in any of the D3D12 modes where MSAA is on.

Were you using the interpolated mode? I think that's just the matching being confused and interpolating it.

**Comment by klorfmorf** [comment]
> > There is another problem which is sometimes visible in D3D12, the half transparent shadow which can be seen in Vulkan with MSAA 2x can also be sometimes seen in any of the D3D12 modes where MSAA is on.
> 
> Were you using the interpolated mode? I think that's just the matching being confused and interpolating it.

Nope. I turned it off in-case it might affect something.

**Comment by DarioSamo** [comment]
So far I've gathered the following information.

- It works fine in NVIDIA D3D12.
- It breaks in NVIDIA Vulkan. It also corrupts RenderDoc really badly.
- It works fine in AMD D3D12.
- It works fine in AMD Vulkan.
- It works fine in NVIDIA Vulkan if I disable Sample Locations (which introduces rendering artifacts if I don't use it elsewhere).

There's no validation layer errors from either D3D12 or Vulkan. This might be a driver issue so I'll try updating and see if behavior changes. The fact RenderDoc breaks so much is making me think I'm triggering a driver bug so far.

**Comment by klorfmorf** [comment]
I'll spin up a Linux machine and take a look, are you using the open source NVIDIA drivers or the proprietary ones?

**Comment by DarioSamo** [comment]
> I'll spin up a Linux machine and take a look, are you using the open source NVIDIA drivers or the proprietary ones?

I'm not, but someone else tested with the proprietary driver and the issue didn't show up.

**Comment by klorfmorf** [comment]
The issue is still there, using NVIDIA driver version `535.171.04`.

**Vulkan: MSAA Off**
![Ubuntu Vulkan MSAA Off](https://github.com/rt64/rt64/assets/169589138/0d24260e-cc9d-4749-bd35-383b56351fe9)

**Vulkan: MSAA 2x**
![Ubuntu Vulkan MSAA 2x](https://github.com/rt64/rt64/assets/169589138/81160bb0-cc2e-42a6-90f0-3faf3477b869)

**Vulkan: MSAA 4x**
![Ubuntu Vulkan MSAA 4x](https://github.com/rt64/rt64/assets/169589138/23c94c40-d5e5-4078-8b97-d383b469cbff)

Though, a beta driver literally just came out as we're speaking so I'll have to try that.

**Comment by DarioSamo** [comment]
Seems pretty consistent across systems and OS at least. I've verified [turning off sample locations](https://github.com/rt64/rt64/blob/051af43bd857fbfa75e174b1f1276c0b5ad869b0/src/render/rt64_raster_shader.cpp#L325) (which introduces other rendering artifacts that it attempts to fix) does fix the issue at least, so it seems entirely localized to the use of that feature in Vulkan.

I've been completely unable to reproduce the issue in D3D12 so far, so if you can get visual proof of that at its original framerate I'm not sure there's much I can investigate there. The fact the only change triggering the problem is a small PSO change implies I'm either misusing something of the feature (unlikely as there's no validation errors) or there's an actual driver bug lying around (and considering it's fine on AMD I'm inclined towards that).

**Comment by DarioSamo** [comment]
> Here's some confirmation for that "ghost shadow" issue (this is under D3D12 on Windows): https://github.com/rt64/rt64/assets/169589138/1aa4ec92-1969-4919-b4ae-ab1c7823c928

Thanks, I'll see if I can reproduce it in a different spot than the one I was trying out.

Here's the particular driver the person who confirmed it was fine on Linux (but broken on Windows) reported back.

![image](https://github.com/rt64/rt64/assets/538504/60952704-2f55-4991-b25c-1004e2097e03)

So far I've still been completely unable to reproduce this in AMD. There's probably something in particular about the ordering of how this game renders that makes the issue not trigger in Zelda, but I find it hard to believe the issues isn't present there and it's just lucky that it doesn't show up.

**Comment by klorfmorf** [comment]
> > Here's some confirmation for that "ghost shadow" issue (this is under D3D12 on Windows): https://github.com/rt64/rt64/assets/169589138/1aa4ec92-1969-4919-b4ae-ab1c7823c928
> 
> Thanks, I'll see if I can reproduce it in a different spot than the one I was trying out.
> 
> Here's the particular driver the person who confirmed it was fine on Linux (but broken on Windows) reported back.
> 
> ![image](https://private-user-images.githubusercontent.com/538504/332506923-60952704-2f55-4991-b25c-1004e2097e03.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYzMTQyNzAsIm5iZiI6MTcxNjMxMzk3MCwicGF0aCI6Ii81Mzg1MDQvMzMyNTA2OTIzLTYwOTUyNzA0LTJmNTUtNDk5MS1iMjVjLTEwMDRlMjA5N2UwMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUyMVQxNzUyNTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kMmEyYmRkMGY4MjE1Y2EyZWEzMmE0NWIzOWNlMzQwZTJkMGJhM2NmMDQ2YjkzM2E2MGE5NTJjZDUzYjVhOWJmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.0sRmLd_jDxdXl_hyhVQOzCjp60e2B9cjZM-7GALFrDY)
> 
> So far I've still been completely unable to reproduce this in AMD. There's probably something in particular about the ordering of how this game renders that makes the issue not trigger in Zelda, but I find it hard to believe the issues isn't present there and it's just lucky that it doesn't show up.

I'll take a look and I'll try to find some oddities with how the game might be rendering decals.

**Comment by DarioSamo** [comment]
More progress on this: it seems NoN fixes it. It seems like it is related to depth clipping, hence why it doesn't break in Zelda. This is a good point to start with at least but it sounds fixable.

**Comment by klorfmorf** [comment]
> > Here's some confirmation for that "ghost shadow" issue (this is under D3D12 on Windows): https://github.com/rt64/rt64/assets/169589138/1aa4ec92-1969-4919-b4ae-ab1c7823c928
> 
> So you're right, when I did go to this spot, I got the exact same behavior as you did, but that so far seems to be actually consistent and unrelated to sample positions so far. As a matter of fact, I can get it to break and disappear here without any MSAA at all.
> 
> * Turning on to LLE mode, the behavior is more consistent and the decal shows up more consistently. This hints that part of the problem is fixed by lowering the precision of the transformation step, which likely gets the shadow and floor closer in coordinates.
> * The issue also consistently goes more away if I multiply the tolerance of the decal (determined by the depth derivative) by some arbitrary factor. Again, depth and dz had much lower precision on real hardware so this could also explain why it's not as visible.
> * In reality, it is also visible on hardware but not as much. On ParaLLEl you can spot the issue a bit up close to the camera when you walk around this area.
> 
> ![image](https://private-user-images.githubusercontent.com/538504/332600608-88854264-efd4-4861-8649-cd3f43fbe0ba.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYzMzI2NDcsIm5iZiI6MTcxNjMzMjM0NywicGF0aCI6Ii81Mzg1MDQvMzMyNjAwNjA4LTg4ODU0MjY0LWVmZDQtNDg2MS04NjQ5LWNkM2Y0M2ZiZTBiYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUyMVQyMjU5MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01NmRlOTliMTZlNDY0ZjkzMWY2YWY1OTNjNTAxYzUyYjFjYWRkNDNjN2RjYTNmYjRlMzEyM2JiNTUxNjBiZmY3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.kPHlApGy2P__gzCiHRwnkS26x2ABtaYE4LXwHJsdDm8)
> 
> So this particular case you're showing of the D3D12 one is exposing a different error that is consistent across both APIs. But, there's more...
> 
> The whole "the decal becomes transparent" thing goes a bit deeper and is only visible in HLE, not in LLE mode most likely because it's exposing some internal behavior about how dz is computed in the shader that causes some of the samples to pass the depth tolerance test and others to not pass it.
> 
> ![image](https://private-user-images.githubusercontent.com/538504/332601096-c775157a-70b1-487f-9d48-8a5fb1917261.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYzMzI2NDcsIm5iZiI6MTcxNjMzMjM0NywicGF0aCI6Ii81Mzg1MDQvMzMyNjAxMDk2LWM3NzUxNTdhLTcwYjEtNDg3Zi05ZDQ4LThhNWZiMTkxNzI2MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUyMVQyMjU5MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05NmI4MTZlZDdiYzBlY2M5OWU4NzM2NTI4YTMxNTVjODE3ZjhlZDU5MDYwYWRlMjgxYzkxOGRlYjUyNWI5YmNiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.KfPhjhoQ8EvnAUkAFpCDXe_E4SEmOyhCp-UNau7zDUU)
> 
> Now here's the interesting part... it doesn't happen on the ubershader!
> 
> ![image](https://private-user-images.githubusercontent.com/538504/332601152-0b30d2b4-c9f3-4721-9657-b86a73dc13be.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYzMzI2NDcsIm5iZiI6MTcxNjMzMjM0NywicGF0aCI6Ii81Mzg1MDQvMzMyNjAxMTUyLTBiMzBkMmI0LWM5ZjMtNDcyMS05NjU3LWI4NmE3M2RjMTNiZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUyMVQyMjU5MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yZmUwZTVkY2U0MzVmYzYxZjUzY2Y0ZTU0NmI0OTA5ZWU5ZmVhMDE5Y2JjYzkxYzQyYmQzM2YxMTY0MTAzOGM0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.1XI-oAWvUJH3FUK6LiUzYIPJHxyWqZ024eU1MiFa9BM)
> 
> This one's a bit harder to explain, I'm gonna have to see if there's any possible compiler differences that could lead to the dz calculation being off for different samples. Perhaps it could lead to the possible reason behind the Vulkan behavior as well, because it also fixed the problem in Vulkan if I use ubershaders there.
> 
> ![image](https://private-user-images.githubusercontent.com/538504/332601374-b4ea03ed-9849-4ab4-b6fb-bb693fb26981.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYzMzI2NDcsIm5iZiI6MTcxNjMzMjM0NywicGF0aCI6Ii81Mzg1MDQvMzMyNjAxMzc0LWI0ZWEwM2VkLTk4NDktNGFiNC1iNmZiLWJiNjkzZmIyNjk4MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUyMVQyMjU5MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mZDRkMmNhZDcxODcyNzk1NjdkNzQ2ZTRiYzY0YjU5YmY1NTBjNGNiZWYyZTM4ZmI3NDQyYjk2YzA0ZjNmYzk3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.m7WoP2GbTNs_qP_K8kNBpj6QVR9jnuR-AfQ3b--x-xU)
> 
> What's even crazier about that one is that is not going through the shader compiler at all, but through specialization constants. DXC and the driver compiler are ending up with the same error! Feel free to try if this toggle also consistently fixes the behavior on your end.
> 
> ![image](https://private-user-images.githubusercontent.com/538504/332601592-cc03a761-6407-4153-8238-9520a6b31b7a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYzMzI2NDcsIm5iZiI6MTcxNjMzMjM0NywicGF0aCI6Ii81Mzg1MDQvMzMyNjAxNTkyLWNjMDNhNzYxLTY0MDctNDE1My04MjM4LTk1MjBhNmIzMWI3YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUyMVQyMjU5MDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kM2NiYmE4YTFjNjU0YzA1OGI5ZDJkZDNiNGVlYTEyMTM2M2FkODJlYmU0ZTE3MDk1YzY0NzQwYTE4NTI0ZTJlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.t0Zc1pOlIEHNuvD38JknJWxw21X-q7Qtv8FWdwhuaRs)
> 
> Seems like I'll have to dig down further but this answers a few things at least. I think at least on your end you might want to review if the game is not being generous enough with placing the decal on the ground for this area at least, as that one might be unavoidable if the ParaLLEl shot is anything to go by.

Great progress! I can confirm it doesn't occur with the Ubershaders only flag. As you said that area has problems on hardware as well, there's a mismatch with the visual floor and collision floor, so the decal isn't flush to the ground.

**Comment by DarioSamo** [comment]
This is consistent across vendors, but basically, if all the other shaders previous to this one use SV_Depth, then the decal won't break. It is completely irrelevant whether the decal is optimized or not, it's the previous ones that matter.

![image](https://github.com/rt64/rt64/assets/538504/73b5f99c-de43-41c8-a4c3-961236b8257e)

Therefore it is completely possible to fix this temporarily if this function always returns true instead:
```
bool ShaderDescription::outputDepth() const {
    bool copyMode = (otherMode.cycleType() == G_CYC_COPY);
    bool depthClampNear = flags.NoN;
    bool depthDecal = (otherMode.zMode() == ZMODE_DEC);
    bool zSourcePrim = (otherMode.zSource() == G_ZS_PRIM);
    return !copyMode && (depthClampNear || depthDecal || zSourcePrim);
}
```
Which as you can see, it always does in NoN mode, hence why the issue never shows up in Zelda.

The issue is consistent across both vendors so it seems I just have an error on my end when understanding what operations are allowed if you seemingly output depth on some shaders and not on others, which breaks the multisampling behavior. So if a fix is needed, then patching the function above is perfectly reasonable: performance is worse in theory but Zelda completely needs it and it's a far heavier game than this I imagine. I mostly have this as an optimization if applicable, but apparently it can cause this bug at the moment.

Feel free to patch the function and confirm if it fixes the MSAA behavior on your end.

**Comment by klorfmorf** [comment]
> This is consistent across vendors, but basically, if all the other shaders previous to this one use SV_Depth, then the decal won't break. It is completely irrelevant whether the decal is optimized or not, it's the previous ones that matter.
> 
> ![image](https://private-user-images.githubusercontent.com/538504/332608480-73b5f99c-de43-41c8-a4c3-961236b8257e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTYzMzU5NTQsIm5iZiI6MTcxNjMzNTY1NCwicGF0aCI6Ii81Mzg1MDQvMzMyNjA4NDgwLTczYjVmOTljLWRlNDMtNDFjOC1hNGMzLTk2MTIzNmI4MjU3ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUyMVQyMzU0MTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04ZWQ4NjA3YzZkY2E1ZjU0ZDQ1YjVjODQ4NDc1MzZmNTQ5N2IzZWJlZTcyNjQxZDBlZTUzYjFjZmUzM2ViNDZkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.4w1dnAFNO66UJ8qH4iEkZAsZ7hsek4W1aEtC5aJ3LKM)
> 
> Therefore it is completely possible to fix this temporarily if this function always returns true instead:
> 
> ```
> bool ShaderDescription::outputDepth() const {
>     bool copyMode = (otherMode.cycleType() == G_CYC_COPY);
>     bool depthClampNear = flags.NoN;
>     bool depthDecal = (otherMode.zMode() == ZMODE_DEC);
>     bool zSourcePrim = (otherMode.zSource() == G_ZS_PRIM);
>     return !copyMode && (depthClampNear || depthDecal || zSourcePrim);
> }
> ```
> 
> Which as you can see, it always does in NoN mode, hence why the issue never shows up in Zelda.
> 
> The issue is consistent across both vendors so it seems I just have an error on my end when understanding what operations are allowed if you seemingly output depth on some shaders and not on others, which breaks the multisampling behavior. So if a fix is needed, then patching the function above is perfectly reasonable: performance is worse in theory but Zelda completely needs it and it's a far heavier game than this I imagine. I mostly have this as an optimization if applicable, but apparently it can cause this bug at the moment.
> 
> Feel free to patch the function and confirm if it fixes the MSAA behavior on your end.

Indeed, that fixes the issue.

**Comment by DarioSamo** [comment]
> Indeed, that fixes the issue.

Alright, as that's a bit of a hack with some undesired consequences elsewhere, I'd roll that back for now. I'm gonna explore an alternate way to fix this. There should be no reason to use depth output on decals now that I think about it, just need to adjust it a bit.

I still think there's some fairly suspicious behavior related to sample locations, but at least this answers the other part of the problem and might lead to fixing it anyway.

**Comment by DarioSamo** [comment]
I've pushed a workaround for this for now. If MSAA is used, it'll forcefully enable depth output in the shader. This seems to fix the issue consistently on everywhere I've tested so far including Vulkan with sample locations. It seems the true reason will escape me for a while longer but I don't think this is impactful enough to matter. It's certainly a better sacrifice than the alternative methods of decal emulation. There is probably a mechanic I need to review about how depth multisampling texture sampling works in conjunction with shaders that can manually choose the depth output.

I leave it up to you if you want to open an issue separately to track about the decals on the problematic area, but I consider this one fixed for now at least. That one I'm fairly sure is due to the higher precision as the error goes away partially on LLE mode (I tested with Mupen here) and it also goes away further the more I drop the precision of the depth buffer in the formula. That's one of those that's a bit of a tough call as it's not flawless in AL/ParaLLEl either.

**Comment by klorfmorf** [comment]
The exact reasoning goes over my head at the moment as I'm not used to modern graphics programming at all, but great work!

>I leave it up to you if you want to open an issue separately to track about the decals on the problematic area, but I consider this one fixed for now at least.

As this is much more of a game issue than anything and considering it's messed up on hardware and LLE mode as well, I'd just leave it in this issue. It only differs from hardware in flickering behaviour (because of the depth precision you mentioned).

>Side note: Do you think it's possible the developers thought they had to add a vertical offset to these even if decals would work correctly at the exact same height? Sounds like a good possibility for a quick patch if that's the case.

I don't think they did, at least not from what I can see. That error there is most probably because the height of the floor in the visual model is far enough apart from that in the collision model.

**Comment by DarioSamo** [comment]
Good news, https://github.com/rt64/rt64/pull/142 seems to address this. Give it a go when you can.
